EventListener/UserContextSubscriber.php

```diff
@@ -124,9 +133,19 @@ public function onKernelResponse(FilterResponseEvent $event)
         }
 
         $response = $event->getResponse();
+        $request = $event->getRequest();
+
         $vary = $response->getVary();
 
-        if ($event->getRequest()->headers->has($this->hashHeader)) {
+        if ($request->headers->has($this->hashHeader)) {
+            // hash has changed, session has most certainly changed, prevent setting incorrect cache
+            if (!is_null($this->hash) && $this->hash !== $request->headers->get($this->hashHeader)) {
+                $response->setClientTtl(0);
+                $response->headers->addCacheControlDirective('no-cache');
```

Comment:
indeed we might need to be more clear with these, override the cache control directives for private no-cache and s-maxage=0 and max-age=0.

also if we use the special custom ttl header we should set it to 0 and make sure its not added again by the header listener.